### PR TITLE
ramips: add support for D-Link DIR-882 R1

### PR DESCRIPTION
**Pull request by pyckle**
The R1 revision is identical to the A1 revision except
- No Config2 Parition, therefore
   - factory partition resized to 64k from 128K
   - Firmware partition offset is 0x50000 not 0x60000
   - Firmware partitions size increased by 64K
- Firmware partition type is "denx,uimage", not "sge,uimage"
- Padding of image creation "uimage-padhdr 96" removed

Installation:
- Older firmware versions: put the factory image on a USB stick, turn on
  the telnet console, and flash using the following cmd
  "fw_updater Linux /mnt/usb_X_X/firmware.bin"

- D-Link FailsafeUI:
  Sign the firmware image with the script here:
  https://pastebin.com/5wGYjTsn
  Then, power down the router, press and hold the reset button, then
  re-plug it. Keep the reset button pressed until the internet LED stops
  flashing, then jack into any lan port and manually assign a static IP
  address in 192.168.0.0/24 other than 192.168.0.0 (e.g. 192.168.0.2)
  and go to http://192.168.0.1

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

### CODE REVIEW

File: target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts:2]: drop dts-v1, it's in mt7621.dtsi now ...

File: target/linux/ramips/image/mt7621.mk [hunk 2]
```diff
@@ -250,6 +250,18 @@ define Device/dlink_dir-8xx-a1
...
+  $(Device/dsa-migration)
+  IMAGE_SIZE := 16064k
+  DEVICE_VENDOR := D-Link
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware
+  KERNEL_INITRAMFS := $$(KERNEL) 
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs |\
+	pad-rootfs | append-metadata | check-size
+  IMAGE/factory.bin := append-kernel | append-rootfs | check-size
+endef
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:263]: Do you expect other similar devices? If not, just put the command into the device's node as long as only one device inherits from it.

**Comment by adschm** [comment]
Please add specifications to your commit message.

Despite, you are settings execute bit on all files, please remove that unless where necessary (01_leds).

File: target/linux/ramips/image/mt7621.mk (hunk 2 above)
Comment (pyckle) [reply target/linux/ramips/image/mt7621.mk:263]: I think that the DIR-878-R1 will have the same partition structure and can reuse these files.

File: target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts (hunk 1 above)
Comment (pyckle) [reply target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts]: Done.

**Comment by pyckle** [comment]
Hi,

Thanks for the timely review (:

I added the specifications to the commit message, removed the execute bits that mistakenly were added, and commented/changed on the other feedback messages.

I am also wondering whether it is worth getting the code to sign the factory image so it can be more easily uploaded onto the router into the build process. Currently it is a bash script here: https://pastebin.com/5wGYjTsn

If I were to do this, where should I place the tool and what language should it be in? Can rewrite this in bash, C, python or perl.

**Comment by pyckle** [comment]
I added the signing process as part of the build. Please let me know what you think.

File: target/linux/ramips/image/mt7621.mk [hunk 3]
```diff
@@ -317,9 +328,20 @@ define Device/dlink_dir-882-a1
   DEVICE_MODEL := DIR-882
   DEVICE_VARIANT := A1
   DEVICE_PACKAGES += kmod-usb3 kmod-usb-ledtrig-usbport
+  IMAGE/factory.bin := append-kernel | append-rootfs | check-size | \
+	  sign-dlink-ru 57c5375741c30ca9ebcb36713db4ba51 \
+	  ab0dff19af8842cdb70a86b4b68d23f7
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:333]: That's added to the wrong device?

Anyway, I'd make the factory image stuff a separate commit ...

File: target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		led-failsafe = &led_power_green;
+		led-running = &led_power_green;
+		led-upgrade = &led_net_orange;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power_orange: power_orange {
+			label = "dir-882-r1:orange:power";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts:20]: I've just pushed a patch to master that removes the model part of these LED labels. With that change, keys and leds are the same as for dir-8xx-a1 as it appears to me. So, the question arises whether we can move more to the DTSI or even create a shared DTSI for dir-8xx-a1 and dir-8xx-r1 ...

File: target/linux/ramips/image/mt7621.mk (hunk 3 above)
Comment (pyckle) [reply target/linux/ramips/image/mt7621.mk]: Whoops! Moved to the right section.

I moved the factory image stuff to a separate commit now too.

File: target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts (hunk 4 above)
Comment (pyckle) [reply target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts]: Nice work (: I adjusted the DTSI/DTS changes to reflect your patch. At this point, the dtsi file between the 8xx-r1 routers and 8xx-a1 routers is related to the partitioning.

I think there are a few options here:
1. Keep things as I pushed them.
2. Make an 8xx-x1 dtsi, have the 8xx-r1.dtsi and 8xx-a1.dtsi import it, and put the partition section there.
3. Only have an 8xx-x1 dtsi and put the partition section in each model's dts file.

What would you like me to do? I would appreciate if somebody throughly review any changes if I refactor the DTSI files. I only have the DIR882-R1 model to test with, and my knowledge of DTSI and DTS files is rudimentary.

Comment (pyckle) [reply target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts]: I created a few dtsi files to reduce duplication. Please let me know what you think.

**Comment by ayavilevich** [comment]
@adschm hey,

Do you know what is preventing this PR from being merged? Please advise on next steps. Having an official support for the DIR-882 R1 would be wonderful.

**Comment by shayx** [comment]
Hey guys, from where can I download a working image for the dir 882 e1?

**Review by yehoshuao — APPROVED** [review]

**Comment by ayavilevich** [comment]
@yehoshuao there is a build at http://luckys.onmypc.net/openwrt/DIR-882-R1/ but I think it is old by now. You can build your own or maybe @pyckle can post an up to date link.

There is also a discussion at https://forum.openwrt.org/t/d-link-dir-882-r1-bootloader-fw-image-verification/74834 . fyi.

**Comment by pyckle** [comment]
@yehoshuao 
A build from this morning is available here: https://mrpickles.fr.to/43402406772d41dbfd1e66e064ea7e5c8a3d8f7e7ed382ee8c29b21f7aeb9512/

$ sha256sum bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-*
5a68c06bc0b64c40d73ca6de3524b57c3c3800850c9283333609072012ab55f8  bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-factory.bin
43402406772d41dbfd1e66e064ea7e5c8a3d8f7e7ed382ee8c29b21f7aeb9512  bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-sysupgrade.bin

**Comment by yeohayon** [comment]
thanks @ayavilevich :)
and thanks a lot @pyckle for your hard work! will test it soon and report back

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
